README.md
# sift.el

[![License GPL 2][badge-license]][LICENSE]
[![Coverage Status](https://coveralls.io/repos/nlamirault/sift.el/badge.png?branch=master)](https://coveralls.io/r/nlamirault/sift.el?branch=master)

Master :
* [![MELPA Stable](https://stable.melpa.org/packages/sift-badge.svg)](https://stable.melpa.org/#/sift)
* [![Circle CI](https://circleci.com/gh/nlamirault/sift.el/tree/master.svg?style=svg)](https://circleci.com/gh/nlamirault/sift.el/tree/master)

Develop:
* [![Melpa Status](https://melpa.org/packages/sift-badge.svg)](https://melpa.org/#/sift)
* [![Circle CI](https://circleci.com/gh/nlamirault/sift.el/tree/develop.svg?style=svg)](https://circleci.com/gh/nlamirault/sift.el/tree/develop)

``sift.el`` allows you to search using [sift][] from inside Emacs.

## Installation

The recommended way to install ``sift`` is via [MELPA][]:

    M-x package-install sift

or [Cask][]:

	(depends-on "sift")


## Usage

<kbd>M-x sift-regexp</kbd> or <kbd>M-x projectile-sift</kbd>



## Development

### Cask

``sift.el`` use [Cask][] for dependencies management. Install it and
retrieve dependencies :

    $ curl -fsSkL https://raw.github.com/cask/cask/master/go | python
    $ export PATH="$HOME/.cask/bin:$PATH"
    $ cask


### Testing

* Launch unit tests from shell

    $ make clean test

* Using [overseer][] :

Keybinding           | Description
---------------------|------------------------------------------------------------
<kbd>C-c , t</kbd>   | launch unit tests from buffer
<kbd>C-c , b</kbd>   | launch unit tests
<kbd>C-c , g</kbd>   | launch unit tests with tag (find, regexp, ...)

* Tips:

If you want to launch a single unit test, add a specify tag :

```lisp
(ert-deftest test-foobar ()
  :tags '(current)
  ```

And launch it using : <kbd>C-c , g</kbd> and specify tag : *current*


## Support / Contribute

See [here](CONTRIBUTING.md)


## Changelog

A changelog is available [here](ChangeLog.md).


## License

See [LICENSE](LICENSE).


## Contact

Nicolas Lamirault <nicolas.lamirault@gmail.com>

[sift.el]: https://github.com/nlamirault/sift.el
[badge-license]: https://img.shields.io/badge/license-GPL_2-green.svg?style=flat
[LICENSE]: https://github.com/nlamirault/sift.el/blob/master/LICENSE

[GNU Emacs]: https://www.gnu.org/software/emacs/
[MELPA]: https://melpa.org/
[Cask]: http://cask.github.io/
[Issue tracker]: https://github.com/nlamirault/sift.el/issues

[overseer]: https://github.com/tonini/overseer.el

[sift]: https://sift-tool.org/



ChangeLog.md
# sift.el ChangeLog

# Version 0.2.0 (01/07/2016)

- Setup Sift search using projectile.
- ``FIX`` CircleCI build

# Version 0.1.0 (11/12/2015)

- Setup unit tests (use [ert-runner][] and [overseer][])
- Init project

[ert-runner]: https://github.com/rejeep/ert-runner.el
[overseer]: https://github.com/tonini/overseer.el


test/test-helper.el
;; test-helper.el --- Test helpers for sift.el

;; Copyright (C) 2015, 2016 Nicolas Lamirault <nicolas.lamirault@gmail.com>

;; Author: Nicolas Lamirault <nicolas.lamirault@gmail.com>
;; Homepage: https://github.com/nlamirault/sift.el

;;; License:

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'ansi)
(require 'cl) ;; http://emacs.stackexchange.com/questions/2864/symbols-function-definition-is-void-cl-macroexpand-all-when-trying-to-instal
(require 'ert)
(require 'f)
(require 'undercover)

(setq debugger-batch-max-lines (+ 50 max-lisp-eval-depth)
      debug-on-error t)

(defvar username (getenv "HOME"))

(defconst sift-testsuite-dir
  (f-parent (f-this-file))
  "The testsuite directory.")

(defconst sift-source-dir
  (f-parent sift-testsuite-dir)
  "The sift.el source directory.")

(defconst sift-sandbox-path
  (f-expand "sandbox" sift-testsuite-dir)
  "The sandbox path for sift.")

(defun cleanup-load-path ()
  "Remove home directory from 'load-path."
  (message (ansi-green "[sift] Cleanup path"))
  (mapc #'(lambda (path)
            (when (string-match (s-concat username "/.emacs.d") path)
              (message (ansi-yellow "Suppression path %s" path))
              (setq load-path (delete path load-path))))
        load-path))

(defun load-unit-tests (path)
  "Load all unit test from PATH."
  (message (ansi-green "[sift] Execute unit tests %s"
                       path))
  (dolist (test-file (or argv (directory-files path t "-test.el$")))
    (load test-file nil t)))


(defun load-library (file)
  "Load current library from FILE."
  (let ((path (s-concat sift-source-dir file)))
    (message (ansi-yellow "[sift] Load library from %s" path))
    (undercover "*.el" (:exclude "*-test.el"))
    (require 'sift path)))


(defmacro with-test-sandbox (&rest body)
  "Evaluate BODY in an empty sandbox directory."
  `(unwind-protect
       (condition-case nil ;ex
           (let ((default-directory sift-source-dir))
             ;; (unless (f-dir? sift-sandbox-path)
             ;;   (f-mkdir sift-sandbox-path))
             (cleanup-load-path)
             (load-library "/sift.el")
             ,@body)
         ;; (f-delete sift-sandbox-path :force)))
         )))
         ;; (error
         ;;  (message (ansi-red "[sift] Error during unit tests : %s" ex))))))

(provide 'test-helper)
;;; test-helper.el ends here


test/sift-version-test.el
;;; sift-version-test.el --- Tests for version information

;; Copyright (C) 2015, 2016 Nicolas Lamirault <nicolas.lamirault@gmail.com>

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
;; 02110-1301, USA.

;;; Commentary:

;;; Code:


(require 'test-helper)

(ert-deftest phpunit-mode-library-version ()
  :tags '(version)
  :expected-result (if (executable-find "cask") :passed :failed)
  (let* ((cask-version (car (process-lines "cask" "version"))))
    (message "sift.el Cask version: %s" cask-version)
    (should (string= "0.3.0" cask-version))))


(provide 'sift-version-test)
;;; sift-version-test.el ends here


projectile-sift.el
;;; projectile-sift.el --- Run a sift with Projectile

;; Copyright (C) 2015, 2016 Nicolas Lamirault <nicolas.lamirault@gmail.com>
;;
;; Author: Nicolas Lamirault <nicolas.lamirault@gmail.com>
;; Version: 0.3.0
;; Keywords : sift projectile
;; Homepage: https://github.com/nlamirault/sift.el

;; Package-Requires: ((sift "0.3.0") (projectile "0.13.0"))

;;; Commentary:

;; Please see README.md for documentation, or read it online at
;; https://github.com/nlamirault/sift.el

;;; License:

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
;; 02110-1301, USA.

;;; Commentary:

;; Usage :

;; M-x projectile-sift

;;; Code:


(require 'projectile)
(require 'sift)


;;;###autoload
(defun projectile-sift (regexp)
  "Run a sift search with `REGEXP' rooted at the current projectile project root."
  (interactive
   (list
    (read-from-minibuffer "Sift search for: " (thing-at-point 'symbol))))
  (if (fboundp 'projectile-project-root)
      (sift-regexp regexp
                   (projectile-project-root)
                   (mapcar (lambda (val) (concat "--not-file-matches=" val))
                           (append projectile-globally-ignored-files
                                   projectile-globally-ignored-directories)))
    (error "Projectile is not available")))


(provide 'projectile-sift)
;;; projectile-sift.el ends here


sift.el
;;; sift.el --- Front-end for sift, a fast and powerful grep alternative

;; Copyright (C) 2015, 2016 Nicolas Lamirault <nicolas.lamirault@gmail.com>
;;
;; Author: Nicolas Lamirault <nicolas.lamirault@gmail.com>
;; Version: 0.3.0
;; Keywords : sift ack pt ag grep search
;; Homepage: https://github.com/nlamirault/sift.el

;;; Commentary:

;; Please see README.md for documentation, or read it online at
;; https://github.com/nlamirault/sift.el

;;; License:

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
;; 02110-1301, USA.

;;; Commentary:

;; Usage :

;; M-x sift-regexp

;;; Code:

(require 'compile)
(require 'grep)
(require 'thingatpt)


;; Customization
;; --------------------------


(defcustom sift-executable
  "sift"
  "Name of the sift executable to use."
  :type 'string
  :group 'sift)


(defcustom sift-arguments
  (list "-I")
  "Default arguments passed to sift."
  :type '(repeat (string))
  :group 'sift)


(defcustom sift-highlight-search t
  "Non-nil means we highlight the current search term in results.
This requires the sift command to support --color-match, which is only in v0.14+"
  :type 'boolean
  :group 'sift)


;; Faces
;; --------------------------


(defface sift-hit-face '((t :inherit compilation-info))
  "Face name to use for sift matches."
  :group 'sift)


(defface sift-match-face '((t :inherit match))
  "Face name to use for sift matches."
  :group 'sift)



;; Mode
;; --------------------------


(defvar sift-search-finished-hook nil
  "Hook run when sift completes a search in a buffer.")

(defun sift/run-finished-hook (buffer how-finished)
  "Run the sift hook to signal that the search has completed."
  (with-current-buffer buffer
    (run-hooks 'sift-search-finished-hook)))

(defvar sift-search-mode-map
  (let ((map (make-sparse-keymap)))
    (set-keymap-parent map compilation-minor-mode-map)
    (define-key map "p" 'compilation-previous-error)
    (define-key map "n" 'compilation-next-error)
    (define-key map "{" 'compilation-previous-file)
    (define-key map "}" 'compilation-next-file)
    (define-key map "k" '(lambda ()
                           (interactive)
                           (let ((kill-buffer-query-functions))
                             (kill-buffer))))
    map)
  "Keymap for sift-search buffers.
`compilation-minor-mode-map' is a cdr of this.")


(define-compilation-mode sift-search-mode "Sift"
  "Platinum searcher results compilation mode"
  (set (make-local-variable 'truncate-lines) t)
  (set (make-local-variable 'compilation-disable-input) t)
  (set (make-local-variable 'tool-bar-map) grep-mode-tool-bar-map)
  (let ((symbol 'compilation-sift)
        (pattern '("^\\([^:\n]+?\\):\\([0-9]+\\):[^0-9]" 1 2)))
    (set (make-local-variable 'compilation-error-regexp-alist) (list symbol))
    (set (make-local-variable 'compilation-error-regexp-alist-alist) (list (cons symbol pattern))))
  (set (make-local-variable 'compilation-error-face) 'sift-hit-face)
  (add-hook 'compilation-filter-hook 'sift-filter nil t))


;; Taken from grep-filter, just changed the color regex.
(defun sift-filter ()
  "Handle match highlighting escape sequences inserted by the ag process.
This function is called from `compilation-filter-hook'."
  (when sift-highlight-search
    (save-excursion
      (forward-line 0)
      (let ((end (point)) beg)
        (goto-char compilation-filter-start)
        (forward-line 0)
        (setq beg (point))
        ;; Only operate on whole lines so we don't get caught with part of an
        ;; escape sequence in one chunk and the rest in another.
        (when (< (point) end)
          (setq end (copy-marker end))
          ;; Highlight ag matches and delete marking sequences.
          (while (re-search-forward "\033\\[30;43m\\(.*?\\)\033\\[[0-9]*m" end 1)
            (replace-match (propertize (match-string 1)
                                       'face nil 'font-lock-face 'sift-match-face)
                           t t))
          ;; Delete all remaining escape sequences
          (goto-char beg)
          (while (re-search-forward "\033\\[[0-9;]*[mK]" end 1)
            (replace-match "" t t)))))))



;; API
;; --------------------------


;;;###autoload
(defun sift-regexp (regexp directory &optional args)
  "Run a sift search with `REGEXP' rooted at `DIRECTORY'.
`ARGS' provides Sift command line arguments."
  (interactive
   (list (read-from-minibuffer "Sift search for: " (thing-at-point 'symbol))
         (read-directory-name "Directory: ")))
  (let ((default-directory directory))
    (compilation-start
     (mapconcat 'identity
                (append (list sift-executable)
                        sift-arguments
                        args
                        '("--color" "-n" "--stats")
                        (list (shell-quote-argument regexp) ".")) " ")
     'sift-search-mode)))


(provide 'sift)
;;; sift.el ends here
